arch/x86/pagetables.c

```diff
@@ -246,28 +249,185 @@ void *vmap_user(void *va, mfn_t mfn, unsigned int order,
                  l1_flags);
 }
 
+static int _vunmap(cr3_t *cr3_ptr, void *va) {
+    pgentry_t *tab;
+
+    if (!va)
+        return false;
+
+    dprintk("%s: va: %p (cr3: 0x%p)\n", __func__, va, cr3_ptr);
+
+    spin_lock(&vmap_lock);
+
+    if (mfn_invalid(cr3_ptr->mfn)) {
+        spin_unlock(&vmap_lock);
+        return -EINVAL;
+    }
+
+    tab = init_map_mfn(cr3_ptr->mfn);
+#if defined(__x86_64__)
+    pml4_t *l4e = l4_table_entry((pml4_t *) tab, va);
+    if (mfn_invalid(l4e->mfn) || !l4e->P) {
+        spin_unlock(&vmap_lock);
+        return -ENOENT;
+    }
+
+    tab = init_map_mfn(l4e->mfn);
+#endif
+    pdpe_t *l3e = l3_table_entry((pdpe_t *) tab, va);
+    if (mfn_invalid(l3e->mfn) || !l3e->P) {
+        spin_unlock(&vmap_lock);
+        return -ENOENT;
+    }
+
+    if (l3e->PS) {
+        set_pgentry(&l3e->entry, MFN_INVALID, PT_NO_FLAGS);
+        spin_unlock(&vmap_lock);
+        return 0;
+    }
+
+    tab = init_map_mfn(l3e->mfn);
+    pde_t *l2e = l2_table_entry((pde_t *) tab, va);
+    if (mfn_invalid(l2e->mfn) || !l2e->P) {
+        spin_unlock(&vmap_lock);
+        return -ENOENT;
+    }
+
+    if (l2e->PS) {
+        set_pgentry(&l2e->entry, MFN_INVALID, PT_NO_FLAGS);
+        spin_unlock(&vmap_lock);
+        return 0;
+    }
+
+    tab = init_map_mfn(l2e->mfn);
+    pte_t *l1e = l1_table_entry((pte_t *) tab, va);
+    if (mfn_invalid(l1e->mfn) || !l1e->P) {
+        spin_unlock(&vmap_lock);
+        return -ENOENT;
+    }
+
+    set_pgentry(&l1e->entry, MFN_INVALID, PT_NO_FLAGS);
+    spin_unlock(&vmap_lock);
+    return 0;
+}
+
+int vunmap_kern(void *va) {
+    return _vunmap(&cr3, va);
+}
+
+int vunmap_user(void *va) {
+    return _vunmap(&user_cr3, va);
+}
+
+#define MMAP_RANGE_SIZE(size, mfn, flags, mmap_flags)                                    \
+    do {                                                                                 \
+        if (likely((mmap_flags) &MMAP_KERNEL))                                           \
+            kmap_##size((mfn), (flags));                                                 \
+        if ((mmap_flags) &MMAP_IDENT)                                                    \
+            vmap_##size(mfn_to_virt((mfn)), (mfn), (flags));                             \
+        if ((mmap_flags) &MMAP_USER)                                                     \
+            vmap_user_##size(mfn_to_virt_user((mfn)), (mfn), (flags | _PAGE_USER));      \
+        if ((mmap_flags) &MMAP_USER_KERNEL)                                              \
+            vmap_##size(mfn_to_virt_user((mfn)), (mfn), (flags & ~_PAGE_USER));          \
+        if ((mmap_flags) &MMAP_KERNEL_MAP)                                               \
+            vmap_##size(mfn_to_virt_map((mfn)), (mfn), (flags));                         \
+    } while (0)
```

Comment:
It is actually by design to support multiple mappings into different areas. Quite often we need to map the same region into VIRT_KERNEL_BASE area and/or VIRT_USER_BASE (or IDENT).

Good point about the return values. The only thing that makes sense here (since the return values are mainly ignored later one) is to `BUG_ON(NULL)` I think.